### PR TITLE
[CDAP-17019] Optimize UI rebuild performance in development mode up to 95.5% + HMR

### PR DESCRIPTION
**Pull request by jennac3**
Optimize performance for incremental UI build in development mode.

Currently there is 15~20s wait for UI to reflect changes in our codebase. This PR optimizes the UI build performance in development mode to speed up CDAP UI developments workflow.

### Performance Improvement
Before - Approximately 15-20s
After - Approximately 0.9-2s

Up to around _90% optimization_
Please note the UI build time heavily depends on the machine.

I added two types of build watch mode.

- Page refresh (`yarn run reload-dev-w`)
Refreshes UI with page refresh, does not retain local state
**Pros**: Stable
**Cons**: 1~2s delay compared to HMR

- Hot module replacement (`yarn run hmr-dev-w`)
Refreshes UI without page refresh, retains local state
**Pros**: Faster, retains local state
**Cons**: Does not support `shouldComponentUpdate`

Reflecting changes without refresh in few seconds

hmr3

### Steps to reproduce:
1. rm -rf node_modules && yarn cache clean && yarn && bower install
2. kill -9 `pgrep -x node`
3. yarn start
4. yarn run hmr-build-w OR yarn run hmr-reload-w

### CODE REVIEW

File: cdap-ui/package.json [hunk 1]
```diff
@@ -73,6 +75,7 @@
     "babel-eslint": "7.2.3",
     "babel-jest": "24.9.0",
     "babel-loader": "8.0.6",
+    "babel-plugin-dynamic-import-node": "^2.3.3",
```
Comment (elfenheart) [inline cdap-ui/package.json]: this doesn't seem to be used

File: cdap-ui/package.json [hunk 2]
```diff
@@ -140,15 +145,18 @@
     "tslint": "5.20.1",
     "tslint-config-prettier": "1.18.0",
     "tslint-plugin-prettier": "2.1.0",
+    "uglifyjs-webpack-plugin": "1.0.1",
     "url-loader": "3.0.0",
     "webpack": "4.41.2",
+    "webpack-bundle-analyzer": "^3.8.0",
```
Comment (elfenheart) [inline cdap-ui/package.json]: I thought this was removed?

File: cdap-ui/package.json (hunk 1 above)
Comment (jennac3) [reply cdap-ui/package.json]: done

File: cdap-ui/package.json (hunk 2 above)
Comment (jennac3) [reply cdap-ui/package.json]: done

File: cdap-ui/webpack.config.cdap.dev.js [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+    proxy: {
+      '/api': {
+        target: 'http://localhost:11011',
+        pathRewrite: { '^/api': '' }
+      }
+    }
+  }
+};
+
+webpackConfig.devtool = 'eval-cheap-module-source-map';
```
Comment (elfenheart) [inline cdap-ui/webpack.config.cdap.dev.js]: why not just set this directly on the webpackConfig object?

File: cdap-ui/package.json [hunk 4]
```diff
@@ -140,6 +144,7 @@
     "tslint": "5.20.1",
     "tslint-config-prettier": "1.18.0",
     "tslint-plugin-prettier": "2.1.0",
+    "uglifyjs-webpack-plugin": "1.0.1",
```
Comment (elfenheart) [inline cdap-ui/package.json]: not being used

File: cdap-ui/webpack.config.cdap.dev.js (hunk 3 above)
Comment (jennac3) [reply cdap-ui/webpack.config.cdap.dev.js]: Resolved

File: cdap-ui/package.json (hunk 4 above)
Comment (jennac3) [reply cdap-ui/package.json]: Resolved

**Review by elfenheart — APPROVED** [review]
:+1: LGTM

**Pull request merged**